packages/itwinui-react/src/core/Tabs/Tabs.tsx

```diff
@@ -512,99 +389,506 @@ export const Tabs = (props: TabsProps) => {
+TabsActions.displayName = 'Tabs.Actions';
 
-      {actions && (
-        <Box className='iui-tabs-actions-wrapper'>
-          <Box className='iui-tabs-actions'>{actions}</Box>
-        </Box>
-      )}
+// ----------------------------------------------------------------------------
+// Tabs.Panel component
+
+type TabsPanelOwnProps = {
+  /**
+   * Value used to associate the panel with a given tab.
+   */
+  value: string;
+};
+
+const TabsPanel = React.forwardRef((props, ref) => {
+  const { value, className, children, ...rest } = props;
+
+  const { activeValue } = useSafeContext(TabsContext);
+
+  return (
+    <Box
+      className={cx('iui-tabs-content', className)}
+      // aria-labelledby={``}
+      role='tabpanel'
+      data-iui-hidden={activeValue !== value ? true : undefined}
+      id={value}
+      ref={ref}
+      {...rest}
+    >
+      {children}
+    </Box>
+  );
+}) as PolymorphicForwardRefComponent<'div', TabsPanelOwnProps>;
+TabsPanel.displayName = 'Tabs.Panel';
+
+// ----------------------------------------------------------------------------
+// Tabs legacy component
+
+type TabsLegacyProps = {
+  /**
+   * Content displayed to the right/bottom of the horizontal/vertical tabs
+   *
+   * If `type = 'pill'`, `actions` is not applicable.
+   */
+  actions?: React.ReactNode[];
+  /**
+   * Elements shown for each tab.
+   * Recommended to pass an array of `Tab` components.
+   */
+  labels: React.ReactNode[];
+  /**
+   * Handler for activating a tab.
+   */
+  onTabSelected?: (index: number) => void;
+  /**
+   * Index of the active tab.
+   */
+  activeIndex?: number;
+  /**
+   * Control whether focusing tabs (using arrow keys) should automatically select them.
+   * Use 'manual' if tab panel content is not preloaded.
+   * @default 'auto'
+   */
+  focusActivationMode?: 'auto' | 'manual';
+  /**
+   * Color of the bar on the active tab.
+   * @default 'blue'
+   */
+  color?: 'blue' | 'green';
+  /**
+   * Custom CSS class name for tabs.
+   */
+  tabsClassName?: string;
+  /**
+   * Custom CSS class name for tab panel.
+   */
+  contentClassName?: string;
+  /**
+   * Custom CSS class name for the tabs wrapper.
+   */
+  wrapperClassName?: string;
+  /**
+   * Content inside the tab panel.
+   */
+  children?: React.ReactNode;
+} & TabsOrientationProps &
+  TabsOverflowProps;
+
+const LegacyTabsComponent = React.forwardRef((props, forwardedRef) => {
+  let actions: Array<React.ReactNode> | undefined;
+  if (props.type !== 'pill' && props.actions) {
+    actions = props.actions;
+    props = { ...props };
+    delete props.actions;
+  }
+
+  const {
+    labels,
+    onTabSelected,
+    focusActivationMode,
+    color,
+    activeIndex,
+    tabsClassName,
+    contentClassName,
+    wrapperClassName,
+    children,
+    ...rest
+  } = props;
+
+  const [currentActiveValue, setCurrentActiveValue] = React.useState<string>();
+
+  // TODO: fix legacy active logic
```

Comment:
this can be fixed using `onActiveChange` and `useControlledState`

<details>
<summary>diff</summary>

```diff
@@ -15,6 +15,7 @@ import {
   useResizeObserver,
   ButtonBase,
   mergeEventHandlers,
+  useControlledState,
 } from '../utils/index.js';
 import { Icon } from '../Icon/Icon.js';
 import type { PolymorphicForwardRefComponent } from '../utils/index.js';
@@ -593,7 +594,7 @@ const LegacyTabsComponent = React.forwardRef((props, forwardedRef) => {
     onTabSelected,
     focusActivationMode,
     color,
-    activeIndex,
+    activeIndex: activeIndexProp,
     tabsClassName,
     contentClassName,
     wrapperClassName,
@@ -601,16 +602,11 @@ const LegacyTabsComponent = React.forwardRef((props, forwardedRef) => {
     ...rest
   } = props;
 
-  const [currentActiveValue, setCurrentActiveValue] = React.useState<string>();
-
-  // TODO: fix legacy active logic
-  const findActiveIndex = () => {
-    return (
-      labels.findIndex(
-        (label, index) => currentActiveValue === `label-${index}`,
-      ) ?? `label-0`
-    );
-  };
+  const [activeIndex, setActiveIndex] = useControlledState(
+    0,
+    activeIndexProp,
+    onTabSelected,
+  );
 
   return (
     <TabsWrapper className={wrapperClassName} {...rest}>
@@ -618,32 +614,23 @@ const LegacyTabsComponent = React.forwardRef((props, forwardedRef) => {
         color={color}
         className={tabsClassName}
         focusActivationMode={focusActivationMode}
-        // activeValue={currentActiveValue}
-        // setActiveValue={(value) => setCurrentActiveValue(value)}
         ref={forwardedRef}
       >
         {labels.map((label, index) => {
-          const tabValue = `label-${index}`;
+          const tabValue = `tab-${index}`;
           return React.isValidElement(label) ? (
             React.cloneElement(label as JSX.Element, {
               value: tabValue,
-              active: activeIndex === index || tabValue === currentActiveValue,
-              onClick: (args: unknown) => {
-                onTabSelected?.(index);
-                setCurrentActiveValue(tabValue);
-                label.props.onClick?.(args);
-              },
+              active: index === activeIndex,
+              onActiveChange: () => setActiveIndex(index),
             })
           ) : (
             <LegacyTab
               key={index}
               value={tabValue}
               label={label}
-              active={activeIndex === index || tabValue === currentActiveValue}
-              onClick={() => {
-                onTabSelected?.(index);
-                setCurrentActiveValue(tabValue);
-              }}
+              active={index === activeIndex}
+              onActiveChange={() => setActiveIndex(index)}
             />
           );
         })}
@@ -652,13 +639,9 @@ const LegacyTabsComponent = React.forwardRef((props, forwardedRef) => {
       {actions && <TabsActions>{actions}</TabsActions>}
 
       {children && (
-        <Box // TODO: use Tabs.Panel
-          className={cx('iui-tabs-content', contentClassName)}
-          role='tabpanel'
-          aria-labelledby={`label-${findActiveIndex()}`}
-        >
+        <TabsPanel value={`tab-${activeIndex}`} className={contentClassName}>
           {children}
-        </Box>
+        </TabsPanel>
       )}
     </TabsWrapper>
   );
@@ -694,6 +677,12 @@ type TabLegacyProps = {
    * This will be automatically set by the parent `Tabs` component.
    */
   active?: boolean;
+  /**
+   * Callback fired when the tab becomes active.
+   *
+   * This will be automatically set by the parent `Tabs` component.
+   */
+  onActiveChange?: () => void;
 };
```

</details>

Reply:
Using your suggestion